[BLOCKER] Codex no toma agents.md: forzar carga determinística + preflight en init.sh (sin romper funcionalidad actual)

Contexto
Tras las últimas actualizaciones, Codex dejó de reconocer las definiciones de agents.md. Sospechas: lectura desde otra ruta/branch, caché estancada o formato no esperado. Necesitamos que siempre cargue el agents.md correcto y que el pipeline falle temprano si no lo hace—sin alterar la funcionalidad existente.

Alcance

No cambiar la lógica de agentes/tareas actual.

Agregar mecanismos idempotentes y reversibles:

Forzar ruta de agents.md.

Limpiar cachés de Codex antes de correr.

Smoke test (sentinel) mínimo para validar carga real.

Preflight con logs útiles (sin secretos).

Bandera para desactivar el preflight si se requiere.

Tareas
A) Actualizar agents.md (sin romper lo existente)

Agregar marca de versión detectable en la primera línea (para verificar en logs):

<!-- AGENTS_MD_VERSION: 2025-09-23T00:00:00Z -->


Usar ISO-8601 y actualizarla automáticamente cuando se edite el archivo.

Agregar agente/tarea sentinel (nombres únicos para no colisionar):

### agent: diag-echo
- description: Diagnóstico: debe imprimir un texto fijo si agents.md fue cargado.
- run:
  - echo "DIAG_OK from agents.md"

### task: diag-echo-task
- agent: diag-echo
- steps:
  - run: echo "DIAG_TASK_OK"


No tocar agentes/tareas existentes.

El sentinel no debe invocarse en producción normal; sólo por preflight.

Validación de formato

Asegurar agents.md en UTF-8 (LF, sin BOM).

No introducir tabs en listas tipo - key: value (2 espacios).

B) Actualizar init.sh con preflight y control de caché (mantener compatibilidad)

Forzar ruta del archivo (respetando una env preexistente si ya la usan):

# Antes de invocar Codex
: "${CODEX_AGENTS_PATH:="$PWD/agents.md"}"
export CODEX_AGENTS_PATH


Limpiar caché de Codex de forma segura e idempotente:

rm -rf ~/.codex/cache 2>/dev/null || true
rm -rf /tmp/codex* 2>/dev/null || true


Preflight con modo bypass

Implementar variable CODEX_PREFLIGHT_SKIP=true para saltar el preflight cuando sea necesario (por ejemplo, en jobs muy restringidos).

Preflight no debe imprimir secretos. Sólo metadata técnica y salidas del sentinel.

Bloque de preflight (insertar antes de cualquier ejecución real de Codex):

# --- Preflight Codex (seguro e idempotente) ---
if [ "${CODEX_PREFLIGHT_SKIP:-false}" != "true" ]; then
  set -euo pipefail

  echo "[INIT] PWD=$(pwd)"
  echo "[INIT] GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD 2>/dev/null || echo 'unknown')"
  if [ -f "$CODEX_AGENTS_PATH" ]; then
    echo "[INIT] agents.md PATH=$CODEX_AGENTS_PATH"
    echo "[INIT] agents.md SHA1=$(sha1sum "$CODEX_AGENTS_PATH" | cut -d' ' -f1)"
    # Show first 1 line to see version marker (safe, no secrets)
    head -n 1 "$CODEX_AGENTS_PATH"
  else
    echo "[ERROR] agents.md no encontrado en $CODEX_AGENTS_PATH"
    exit 1
  fi

  # Imprimir configuración cargada (sin secretos) y chequear que mencione agents.md
  if command -v codex >/dev/null 2>&1; then
    echo "[INIT] codex --debug --dry-run --print-config (primeros 200 lines)"
    codex --debug --dry-run --print-config | sed -n '1,200p' || true
  else
    echo "[ERROR] 'codex' no está en PATH"
    exit 1
  fi

  # Smoke test: debe existir y ejecutarse diag-echo-task si cargó agents.md
  echo "[INIT] Ejecutando sentinel diag-echo-task"
  if codex run diag-echo-task --debug 2>&1 | tee /tmp/codex_diag.log; then
    if ! grep -qE 'DIAG_OK from agents.md|DIAG_TASK_OK' /tmp/codex_diag.log; then
      echo "[ERROR] Sentinel ejecutado pero no se detectó salida esperada (¿no cargó agents.md?)"
      exit 1
    fi
    echo "[INIT] Sentinel OK"
  else
    echo "[ERROR] No se pudo ejecutar diag-echo-task (posible no-carga de agents.md)"
    exit 1
  fi
else
  echo "[INIT] CODEX_PREFLIGHT_SKIP=true → Saltando preflight"
fi
# --- Fin preflight ---


Mantener funcionalidades actuales

No modificar los comandos existentes que invocan tareas reales de Codex.

El preflight corre antes y falla rápido si agents.md no fue cargado.

Si se requiere desactivarlo en algún job, setear CODEX_PREFLIGHT_SKIP=true.

(Opcional) Pin de versión de Codex

Agregar soporte para fijar versión sólo si la env está presente (no imponer pin por defecto):

if [ -n "${CODEX_VERSION:-}" ]; then
  echo "[INIT] Usando Codex versión fijada: $CODEX_VERSION"
  # (Place holder: lógica de selección/descarga si aplica a su entorno)
fi

Criterios de aceptación

codex --debug --dry-run --print-config evidencia la ruta correcta de agents.md y aparece la marca AGENTS_MD_VERSION en logs.

codex run diag-echo-task --debug imprime DIAG_OK from agents.md y DIAG_TASK_OK.

El pipeline falla temprano si agents.md no se cargó, con mensaje claro.

Seteando CODEX_PREFLIGHT_SKIP=true el preflight se omite y el resto funciona igual que antes.

No se alteran agentes ni tareas existentes; todo lo previo sigue ejecutando igual.

Cambios son idempotentes y no filtran secretos en logs.

Notas de implementación

Conservar finales de línea LF en agents.md y init.sh.

Evitar set -x para no volcar variables sensibles. El preflight no imprime secrets.

Si hay múltiples agents.md (monorepo), CODEX_AGENTS_PATH debe apuntar al correcto (ruta absoluta o ${PWD} controlado).

Labels: bug, high priority, ci, codex, automation
Proyecto/Milestone: (completar según repo)
Impacto: Evita regresiones silenciosas y reduce tiempos muertos de debugging.

--- a/agents.md
+++ b/agents.md
@@ -1,3 +1,4 @@
+<!-- AGENTS_MD_VERSION: 2025-09-24T00:51:43Z -->
 # agents.md
 
 ## 📘 Descripción General
@@ -262,3 +263,19 @@ Esto indicará al agente que debe buscar todos los issues en estado "Todo" y eje
 ### 🔹 Recomendaciones adicionales
 - No deben utilizarse comandos ambiguos como solo `"refinar"` o `"trabajar"` sin contexto, ya que el agente puede solicitar información adicional.
 - Si se desea refinar o trabajar un issue puntual, se debe indicar el número del issue de forma explícita.
+
+---
+
+## 🛡️ Sentinel de diagnóstico para preflight
+
+Estos elementos permiten validar rápidamente que el archivo `agents.md` correcto fue cargado por Codex sin interferir con el flujo normal.
+
+### agent: diag-echo
+- description: Diagnóstico: debe imprimir un texto fijo si agents.md fue cargado.
+- run:
+  - echo "DIAG_OK from agents.md"
+
+### task: diag-echo-task
+- agent: diag-echo
+- steps:
+  - run: echo "DIAG_TASK_OK"

--- a/init.sh
+++ b/init.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
+
 # Instalando Android SDK...
 echo "📦 Instalando Android SDK..."
 
@@ -97,3 +99,66 @@ else
     fi
   fi
 fi
+
+# --- Preflight Codex (seguro e idempotente) ---
+pushd "$SCRIPT_DIR" >/dev/null
+
+: "${CODEX_AGENTS_PATH:="$PWD/agents.md"}"
+export CODEX_AGENTS_PATH
+
+echo "[INIT] Limpiando caché de Codex..."
+rm -rf ~/.codex/cache 2>/dev/null || true
+rm -rf /tmp/codex* 2>/dev/null || true
+
+if [ -n "${CODEX_VERSION:-}" ]; then
+  echo "[INIT] Usando Codex versión fijada: $CODEX_VERSION"
+  # (Place holder: lógica de selección/descarga si aplica a su entorno)
+fi
+
+if [ "${CODEX_PREFLIGHT_SKIP:-false}" != "true" ]; then
+  (
+    set -euo pipefail
+
+    echo "[INIT] PWD=$(pwd)"
+    echo "[INIT] GIT_BRANCH=$(git rev-parse --abbrev-ref HEAD 2>/dev/null || echo 'unknown')"
+    if [ -f "$CODEX_AGENTS_PATH" ]; then
+      echo "[INIT] agents.md PATH=$CODEX_AGENTS_PATH"
+      echo "[INIT] agents.md SHA1=$(sha1sum "$CODEX_AGENTS_PATH" | cut -d' ' -f1)"
+      # Show first 1 line to see version marker (safe, no secrets)
+      head -n 1 "$CODEX_AGENTS_PATH"
+    else
+      echo "[ERROR] agents.md no encontrado en $CODEX_AGENTS_PATH"
+      exit 1
+    fi
+
+    if command -v codex >/dev/null 2>&1; then
+      echo "[INIT] codex --debug --dry-run --print-config (primeros 200 lines)"
+      codex --debug --dry-run --print-config | sed -n '1,200p' || true
+    else
+      echo "[ERROR] 'codex' no está en PATH"
+      exit 1
+    fi
+
+    echo "[INIT] Ejecutando sentinel diag-echo-task"
+    if codex run diag-echo-task --debug 2>&1 | tee /tmp/codex_diag.log; then
+      if ! grep -qE 'DIAG_OK from agents.md|DIAG_TASK_OK' /tmp/codex_diag.log; then
+        echo "[ERROR] Sentinel ejecutado pero no se detectó salida esperada (¿no cargó agents.md?)"
+        exit 1
+      fi
+      echo "[INIT] Sentinel OK"
+    else
+      echo "[ERROR] No se pudo ejecutar diag-echo-task (posible no-carga de agents.md)"
+      exit 1
+    fi
+  )
+  preflight_status=$?
+  popd >/dev/null
+  if [ "$preflight_status" -ne 0 ]; then
+    exit "$preflight_status"
+  fi
+else
+  echo "[INIT] CODEX_PREFLIGHT_SKIP=true → Saltando preflight"
+  popd >/dev/null
+fi
+
+# --- Fin preflight ---

